### ISSUE TITLE
Should addFunction grow the table less frequently?

### ISSUE THREAD

**Issue by kg** [emscripten-core/emscripten#17891]
Right now addFunction grows the wasm table by 1 every time it adds a new function. This seems like it could be a performance issue, since growing tables could invalidate jitcode by changing the range of valid indices, etc. Would it make sense to grow it in chunks, say 256 items at a time? It doesn't seem like the cost of using an extra say (let's be conservative here) `(32 * 255)` bytes of memory would be a problem. Is it a known thing that this isn't an issue?

I've been jitting wasm functions and importing them via addFunction and seeing mysterious perf issues, so I've been trying to track down the causes and this seemed like a possible factor.

**Comment by sbc100** [comment]
Sounds reasonable.  We could even use some kind of quadratic increment.   Seems like relatively small change to `getEmptyTableSlot`.

Thats very interesting that you are jitting wasm functions..   out of interest are you creating an entirely new module each time or using the new WebAssembly.Function from https://github.com/WebAssembly/js-types/blob/main/proposals/js-types/Overview.md?

**Comment by kg** [comment]
I'm jitting modules right now similar to what addFunction does, and may try to cluster them together into larger modules since it seems like having too many small modules kicking around probably would create problems in various parts of browser runtimes. I still have to investigate more where all the performance pits are. Is WebAssembly.Function actually available in a near timeframe in browsers? I hadn't realized it was an option, should definitely investigate it.

FWIW I started growing the table manually by 256 and filling it myself, and didn't notice a big performance improvement so I don't think the problem was grow. (It looks like the set operation may be what actually has the invalidation overhead, from looking at the v8 source code - but it has a few different paths it can take depending on the value you're inserting.)

Growing in larger steps may still be worth it for workloads where adding happens more frequently, though.

**Comment by martenrichter** [comment]
I looked in the chromium profiler for the [jupyterlite xeus kernel](https://github.com/jupyterlite/xeus). 
At longer periods, it is most often inside the emscripten grow function (between extracts). 
According to the data, it is most often inside `getEmptyTableSlot` and `addFunction`, growing the table by one.
Is there a way one may override this function? (For testing?)

**Comment by hoodmane** [comment]
But another thing that works is if you `./emsdk install 4.0.3` then you can go into `upstream/emscripten/src/lib/libaddfunction.js` and change it there. There is no compilation step for the js runtime libraries, if you change it in place the next time you link something your change will get picked up.

**Comment by martenrichter** [comment]
Okay, thanks. So there is no way to change by passing something to the Module, I guess? 
Then I will set up a development container for the Zeus kernel and see if I can benchmark it and if my hypothesis that it is the hot path is true.

**Comment by martenrichter** [comment]
Ok here are some statistics for loading the jupyter-xeus-kernel:
```
Grow by 1
15482 720.5000000093132 
15786 758.800000006333 
15972 726.8999999854714 
Grow by 100
15900 92.40000000223517
15900 90.40000001341105
Grow by 1000
16000 40.399999998509884
Grow by Wasmlength
28324 14.900000000372529
```
It uses different strategies, growing by 1, 100, 1000 and by current `wasmTable.length`.
The first table shows the number of entries the table grows by. (so different values depending on granularity.
The savings for the kernel startup are not huge, but it is almost a second.
The best performance is achieved by growing by current `wasmTable.length`, an exponential scaling strategy.

I will make a PR for this bit of code.

**Comment by sbc100** [comment]
@martenrichter what is you use case for `addFunction` is it libdylink.js, or are you calling it directly for some other reason?

**Comment by martenrichter** [comment]
I do not know. I am currently using jupyter lite for applets having interactive lectures. 
I have seen that this part of emscripten seamed to be a bottleneck from the profiles and decided to make a patch.
It happens during loading of the many binary parts of the python packages, so I think it could be dynlink, but I do not know.

**Comment by sbc100** [comment]
That sounds like the issue is absolutely dynamic linking then.  I don't think there are likely be any other calls the addFunciton so lets assume they are all a consequence of dynamic linking.

**Comment by martenrichter** [comment]
Probably, I do not know if some package do some dynamic compiling afterwards, but it could not be in masses.

**Comment by sbc100** [comment]
I think the likely in that case that the cause (in your case) is the call to `addFunction` which occurs during `updateGOT`.  It will be called once for each function that gets exported from a given DLL:  https://github.com/emscripten-core/emscripten/blob/558635efe7b0893e45758ff2c22a6a7a7572511a/src/lib/libdylink.js#L230

If you have profiling traces from your application you could confirm this by verifying that the call to `addFunction` is coming from `updateGOT`.

**Comment by martenrichter** [comment]
Yes, I have just run a trace and it is coming from updateGOT which is called by relocateExports and postInstantiation.
